### ISSUE TITLE
torch.export.unflatten loses node.meta['val']

### ISSUE THREAD

**Issue by titaiwangms** [pytorch/pytorch#116670]
### 🐛 Describe the bug

`torch.export.unflatten` unflattens graph to keep the original hierarchy of nn.Module. However, the generated `call_module` node doesn't contain meta['val'] information that the flattened graph does.

```python
import torch

class InnerModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.relu = torch.nn.ReLU()

    def forward(self, x):
        return self.relu(x)

class TestModule(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.inner_module = InnerModule()

    def forward(self, x, y):
        out = x + y
        out = self.inner_module(out)
        out = out + x
        out = self.inner_module.relu(out)
        return out

exported_program = torch.export.export(
    TestModule(), args=(torch.randn(3), torch.randn(3))
)

for node in exported_program.graph.nodes:
    if node.op != "output":
        try:
            print(node.meta["val"])
        except KeyError:
            print("ExportedProgram lacks of val in meta")

unflattened_module = torch.export.unflatten(exported_program)
for node in unflattened_module.graph.nodes:
    if node.op != "output":
        try:
            print(node.meta["val"])
        except KeyError:
            print("UnflattenedModule lacks of val in meta")
```

Output:
```bash
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
FakeTensor(..., size=(3,))
UnflattenedModule lacks of val in meta
FakeTensor(..., size=(3,))
UnflattenedModule lacks of val in meta
```

### Versions

Collecting environment information...
PyTorch version: 2.2.0a0+git684ce1b
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: CBL-Mariner/Linux (x86_64)
GCC version: (GCC) 11.2.0
Clang version: 12.0.1
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.135.1-2.cm2-x86_64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             32
On-line CPU(s) list:                0-31
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          1
Stepping:                           7
BogoMIPS:                           5187.81
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology cpuid pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti tpr_shadow vnmi ept vpid fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap clflushopt avx512cd avx512bw avx512vl xsaveopt xsavec xsaves md_clear
Virtualization:                     VT-x
Hypervisor vendor:                  Microsoft
Virtualization type:                full
L1d cache:                          512 KiB (16 instances)
L1i cache:                          512 KiB (16 instances)
L2 cache:                           16 MiB (16 instances)
L3 cache:                           35.8 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-31
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        KVM: Mitigation: VMX disabled
Vulnerability L1tf:                 Mitigation; PTE Inversion; VMX conditional cache flushes, SMT vulnerable
Vulnerability Mds:                  Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Meltdown:             Mitigation; PTI
Vulnerability Mmio stale data:      Vulnerable: Clear CPU buffers attempted, no microcode; SMT Host state unknown
Vulnerability Retbleed:             Vulnerable
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Vulnerable
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, STIBP disabled, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; Clear CPU buffers; SMT Host state unknown

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-coding==1.3.3
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] mypy==1.7.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.0
[pip3] onnx==1.15.0
[pip3] onnxruntime==1.16.3
[pip3] onnxscript==0.1.0.dev20231222
[pip3] optree==0.10.0
[pip3] pytorch-triton==2.1.0+6e4932cda8
[pip3] torch==2.3.0a0+git95a86ed
[pip3] torchaudio==2.2.0.dev20231120+cpu
[pip3] torchvision==0.17.0.dev20231120+cpu
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-include               2023.1.0         h06a4308_46344  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.8           py311h5eee18b_0  
[conda] mkl_random                1.2.4           py311hdb19cb5_0  
[conda] numpy                     1.26.0          py311h08b1b3b_0  
[conda] numpy-base                1.26.0          py311hf175353_0  
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch-triton            2.1.0+6e4932cda8          pypi_0    pypi
[conda] torch                     2.3.0a0+git95a86ed           dev_0    <develop>
[conda] torchaudio                2.2.0.dev20231120+cpu          pypi_0    pypi
[conda] torchfix                  0.2.0                    pypi_0    pypi
[conda] torchvision               0.17.0.dev20231120+cpu          pypi_0    pypi

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @angelayi @suo @ydwu4

**Comment by angelayi** [comment]
Do you mind specifying why you want the "val" metadata on those nodes too? Technically the unflattened module is just a generic eager pytorch module, so we don't guarantee specific metadata, but we're curious to know your use case and possibly what other metadata you might want.

**Comment by titaiwangms** [comment]
Hi @angelayi Happy New Year!
It's the shape and type information I am looking for. I am working on ONNX converter, and typically we work on torch.fx.GraphModule to ONNX graph. The node with `call_function` comes with meta['val'] to allow us to follow shape and type in torch model, which si great. However, in GraphModule, it's flattened, and don't remain the same hierachy, so I came across to find unflatten API which can get me a hierachy module. So, specifically, I want the symbolic output expectation of the module node.

**Comment by titaiwangms** [comment]
Looks like it's resolved in main branch.

**Comment by thiagocrepaldi** [comment]
@angelayi could you help us with this?

**Comment by titaiwangms** [comment]
Hi @angelayi this should be fixed by #118138?

**Comment by angelayi** [comment]
yup!
